crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,176 @@
+use rustpython_parser::ast::{Constant, Expr, ExprConstant, ExprSlice, Ranged};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for indexed access to lists, strings, tuples, bytes, and comprehensions
+/// using a type other than an integer or slice.
+///
+/// ## Why is this bad?
+/// Only integers or slices can be used as indices to these types. Using
+/// other types will result in a `TypeError`.
+///
+/// ## Example
+/// ```python
+/// var = [1, 2, 3]["x"]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// var = [1, 2, 3][0]
+/// ```
+#[violation]
+pub struct InvalidIndexType {
+    var_type: String,
+    idx_type: String,
+    slice_bound: bool,
+}
+
+impl Violation for InvalidIndexType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidIndexType {
+            var_type, idx_type, ..
+        } = self;
+        if self.slice_bound {
+            format!("Slice in indexed access to type `{var_type}` uses type `{idx_type}` as bound instead of an integer.")
+        } else {
+            format!(
+                "Indexed access to type `{var_type}` uses type `{idx_type}` instead of an integer or slice."
+            )
+        }
+    }
+}
+
+/// RUF015
+/// Expects components of a `Subscript` expression
```

Comment:
One thing you can do here (which is newer so not all rules are written this way):

```rust
pub(crate) fn invalid_index_type<'a>(checker: &mut Checker, expr: &'a ast::ExprSubscript) { ... }
```

That ensures that we only pass in an expression of the correct type, as opposed to passing in two expressions that are _assumed_ to be fields of some specific expression type.

You might have to change `ast/mod.rs` to use:

```rust
Expr::Subscript(subscript @ ast::ExprSubscript {
    value,
    slice,
    ctx,
    range: _,
}) => { ... }
```